File: Programmers/프린터.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    queue<pair<int, int>> q;
    priority_queue<int> pq;
    for (int i = 0; i < priorities.size(); i++) {
        q.push(make_pair(i, priorities[i]));
        pq.push(priorities[i]);
    }
    
    while (!q.empty()) {
        int index = q.front().first;
        int priority = q.front().second;
        q.pop();
        if (pq.top()== priority) {
            answer++;
            if (index == location) {
                return answer;
            }
            pq.pop();
        } else {
            q.push(make_pair(index, priority));
        }
    }
    
    return answer;
}

File: Programmers/프린터.swift
import Foundation

func solution(_ priorities:[Int], _ location:Int) -> Int {
  var indexArray = [Int]()
  var queue = priorities
  var count:Int = 0
  priorities.forEach { _ in
    indexArray.append(count)
    count += 1
  }
  
  var result = 0
  while(true) {
    let front = queue.first!
    
    let highPriority = queue.filter {
      $0 > front
    }
    if highPriority.isEmpty {
      result += 1
      queue.removeFirst()
      let index = indexArray.removeFirst()
      if index == location {
        break
      }
    } else {
      let item = queue.removeFirst()
      queue.append(item)
      let index = indexArray.removeFirst()
      indexArray.append(index)
      
    }
  }
  
  return result
}
